applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -92,63 +116,180 @@ static bool is_gnss_activated(void)
 	return false;
 }
 
-static void gnss_status_notify(int status)
+static void gnss_status_notifier_wk(struct k_work *work)
+{
+	gnss_status = (enum gnss_status)k_fifo_get(&gnss_status_fifo, K_NO_WAIT);
+	rsp_send("\r\n#XGPS: 1,%d\r\n", gnss_status);
+}
+
+static void gnss_status_set(enum gnss_status status)
 {
-	if (run_type == RUN_TYPE_AGPS) {
-		rsp_send("\r\n#XAGPS: 1,%d\r\n", status);
-	} else if (run_type == RUN_TYPE_PGPS) {
-		rsp_send("\r\n#XPGPS: 1,%d\r\n", status);
+	/* A notification is sent when the GNSS status is updated.
+	 * However this gets called in interrupt context (from gnss_event_handler()),
+	 * from which UART messages cannot be sent. This task is thus delegated
+	 * to a worker, to which the the new GNSS status is passed.
+	 * A FIFO rather than a single variable is used to not miss GNSS status changes
+	 * in case they happen twice (or more) before the worker had the time to process them.
+	 */
+	const int put_ret = k_fifo_alloc_put(&gnss_status_fifo, (void *)status);
+
+	if (put_ret == 0) {
+		k_work_submit_to_queue(&slm_work_q, &gnss_status_notifier);
```

Comment:
This does not actually launch another worker if the item is already being processed, ran into this with UART implementation. Not likely to happen, but if we want to get all the notifications, you should unload them all in gnss_status_notifier_wk.

Reply:
If I recall correctly, this failed in scenario where there was something running, a second workqueue submit was scheduled and the third one failed. I did not observe any failures when the queue was emptied in while loop while processing UART, but that does not mean that there would not be scenarios that are problematic.

Anyway, I do not think that these notifications come often enough that they would really be lost with your approach either.